### PR TITLE
INTLY-9458 - Add alert for abnormal CSVs

### PR DESCRIPTION
**Pull request by HubertStefanski**
# Description
Adds a new warning alert for abnormal operator CSVs

## JIRA
https://issues.redhat.com/browse/INTLY-9458

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist
- [ ] This change requires a documentation update 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added a test case that will be used to verify my changes 
- [ ] Verified independently on a cluster by reviewer

### CODE REVIEW

File: pkg/products/monitoring/prometheusRules.go [hunk 1]
```diff
@@ -292,6 +292,22 @@ func (r *Reconciler) newAlertsReconciler() resources.AlertReconciler {
...
+				AlertName: "install-upgrade-alerts",
+				Namespace: r.Config.GetOperatorNamespace(),
+				GroupName: "general.rules",
+				Rules: []monitoringv1.Rule{
+					{
+						Alert: "RHMICSVAbnormal",
+						Annotations: map[string]string{
+							"message": "An abnormal CSV was found, possibly preventing an operator from installing/upgrading",
+						},
+						Expr:   intstr.FromString(fmt.Sprintf("csv_abnormal{reason='RequirementsNotMet',phase='Pending'} or csv_abnormal{reason='RequirementsNotMet',phase='Failed'}")),
```
Comment (HubertStefanski) [inline pkg/products/monitoring/prometheusRules.go:305]: Should we consider both Pending and Failing, or should this be just for Failing CSVs?

File: pkg/products/monitoring/prometheusRules.go [hunk 2]
```diff
@@ -292,6 +292,22 @@ func (r *Reconciler) newAlertsReconciler() resources.AlertReconciler {
...
+						For:    "15m",
```
Comment (HubertStefanski) [inline pkg/products/monitoring/prometheusRules.go:306]: Leaving this at 15 minutes, adhering to the other warning severity alerts

File: pkg/products/monitoring/prometheusRules.go [hunk 3]
```diff
@@ -292,6 +292,22 @@ func (r *Reconciler) newAlertsReconciler() resources.AlertReconciler {
...
 			},
+			{
+				AlertName: "install-upgrade-alerts",
+				Namespace: r.Config.GetOperatorNamespace(),
+				GroupName: "general.rules",
+				Rules: []monitoringv1.Rule{
+					{
+						Alert: "RHMICSVAbnormal",
+						Annotations: map[string]string{
+							"message": "An abnormal CSV was found, possibly preventing an operator from installing/upgrading",
```
Comment (HubertStefanski) [inline pkg/products/monitoring/prometheusRules.go:303]: Don't know if this message conveys the information about what's happening sufficiently, thoughts?

File: pkg/products/monitoring/prometheusRules.go [hunk 4]
```diff
@@ -292,6 +292,22 @@ func (r *Reconciler) newAlertsReconciler() resources.AlertReconciler {
 					},
 				},
 			},
+			{
+				AlertName: "install-upgrade-alerts",
```
Comment (HubertStefanski) [inline pkg/products/monitoring/prometheusRules.go:296]: Is this an appropriate name? or perhaps this should be under a different, already existing alert?

**Review by HubertStefanski — COMMENTED** [review]
Few questions for reviewrs, looking for feedback

**Comment by obrienrobert** [comment]
@HubertStefanski new alert may also have to be added to the `alerts_exist.go` test file.

**Comment by HubertStefanski** [comment]
//cc @alanmoran @pepedocs @mciccone @david-martin

File: pkg/products/monitoring/prometheusRules.go (hunk 1 above)
Comment (obrienrobert) [reply pkg/products/monitoring/prometheusRules.go:305]: It's mentioned in the in the Jira description that operators most often get stuck in either phase, but I'm not sure how this translates to the RHMI operator, and I'm not that familiar with upgrades TBH.

If sticking with both Pending and Failed, could the expression be shortened to something like: `csv_abnormal{reason='RequirementsNotMet',phase='Pending|Failed'}` I wonder?

Comment (david-martin) [reply pkg/products/monitoring/prometheusRules.go:305]: `phase=~'Pending|Failed'` should do the trick

File: pkg/products/monitoring/prometheusRules.go (hunk 3 above)
Comment (david-martin) [reply pkg/products/monitoring/prometheusRules.go:303]: How about including some of the labels in the message?

e.g. `RequirementsNotMet for CSV '{{$labels.name}}' in namespace '{{$labels.namespace}}'. Phase is {{$labels.phase}}`

File: pkg/products/monitoring/prometheusRules.go [hunk 5]
```diff
@@ -292,6 +292,22 @@ func (r *Reconciler) newAlertsReconciler() resources.AlertReconciler {
...
+				Namespace: r.Config.GetOperatorNamespace(),
+				GroupName: "general.rules",
+				Rules: []monitoringv1.Rule{
+					{
+						Alert: "RHMICSVAbnormal",
```
Comment (david-martin) [inline pkg/products/monitoring/prometheusRules.go:301]: Lets call it `RHMICSVRequirementsNotMet`

File: pkg/products/monitoring/prometheusRules.go (hunk 2 above)
Comment (david-martin) [reply pkg/products/monitoring/prometheusRules.go:306]: 👍

File: pkg/products/monitoring/prometheusRules.go (hunk 4 above)
Comment (david-martin) [reply pkg/products/monitoring/prometheusRules.go:296]: This seems ok

File: pkg/products/monitoring/prometheusRules.go (hunk 1 above)
Comment (obrienrobert) [reply pkg/products/monitoring/prometheusRules.go:305]: Do we need to limit the expression to RHMI related namespaces?

Comment (obrienrobert) [reply pkg/products/monitoring/prometheusRules.go:305]: Do we need to limit the expression to `RHMI` related namespaces?

Comment (obrienrobert) [reply pkg/products/monitoring/prometheusRules.go:305]: Do we need to limit the expression to `RHMI` related namespaces?

Comment (HubertStefanski) [reply pkg/products/monitoring/prometheusRules.go:305]: I'll try the shortened version again, I had it in before but it wasn't giving me the results I expected when testing, I might have been missing the `~` though, which might be the culprit, I'll also limit the namespaces to `rhmi`, the cluster prometheus already has a similar query that watches all (redhat)namespaces, also we don't want to get alerts for non-managed products so it makes sense to limit it to our namespaces

File: pkg/products/monitoring/prometheusRules.go (hunk 5 above)
Comment (HubertStefanski) [reply pkg/products/monitoring/prometheusRules.go:301]: Noted, will push the change soon

**Comment by pepedocs** [comment]
@HubertStefanski When possible can you please point me to the SOP for this alert? Looks fine to me when all the requests above are done (limit to rhmi ns).

**Comment by Rajagopalan-Ranganathan** [comment]
Verified that the csv_abnormal metric is propagated to the middleware monitoring.

**Comment by Rajagopalan-Ranganathan** [comment]
****Manual Testing steps:**

**Pre-requisite:****

OSD cluster with RHMI installed

Better to run the integreatly operator locally, so that it can be stopped easily. 

**Step 1:  Editing a CSV file**

Edit the CSV file to cause a failure. One easy way is edit the
"image" key and point it to a non-existent image path or version

> Through Command line:

To get the list of all CSVs in all namespaces
$oc get csv --all-namespaces 

Edit the CSV
oc edit  csv <CSV name>

> Through the web-console
  
  Login to the OSD cluster as kubeadmin

> Choose a project of your choice (home --> project)
    Ex: redhat-rhmi-cloud-resources-operator
> Using the search, search for the resource csv (home --> Search)
> Select the CSV resource and choose "yaml"
> Edit the yaml as said before to cause a failure. Save it.

Wait for a couple minutes for the changes to take effect

**Step 2: Verify csv_abnormal alert in cluster prometheus.**

Select project: Openshift-monitoring

Then, Networking --> routes , navigate to the "prometheus-k8s" location 

Use your kubeadmin credentials if requested. ( login with openshift)

Execute the query ```csv_abnormal``` and see the results. It should show the metrics
for the CSV that was modified earlier.

****Step 3:** Verify csv_abnormal in  "redhat-rhmi-middleware-monitoring-operator"**

Select project:  "redhat-rhmi-middleware-monitoring-operator"

Then, Networking --> routes , navigate to the "prometheus-route" location 

Use your kubeadmin crdentials if requested. ( login with openshift)

Execute the query ```csv_abnormal``` and see the results. It should show the metrics
for the CSV that was modified earlier.

**Pull request merged**